### ISSUE TITLE
Should there be flexibility in when readyState is updated when waiting for a key?

### ISSUE THREAD

**Issue by ddorwin** [w3c/encrypted-media#339]
Similar to #336, browsers may update `readyState` at different times. For example, Chrome currently waits three seconds before pausing playback due to video underflow to avoid glitching when the decoder falls slightly behind. The easiest and most consistent thing for Chrome to do would be to use the same behavior for missing key(s).

However, the Wait for Key algorithm currently requires that the `readyState` be updated before the `waitingforkey` event is fired (because it is fired by a task).

Do we want to be this prescriptive or should we provide more implementation flexibility? The suggestion in https://github.com/w3c/encrypted-media/issues/336#issuecomment-254346323 might also provide such flexibility for this case.

@foolip @cpearce

**Comment by foolip** [comment]
> Chrome currently waits three seconds before pausing playback due to video underflow to avoid glitching when the decoder falls slightly behind

I don't know the context of this, is this about the behavior when running out of data over the network for regular media element playback, just with MSE, or only MSE+EME? Whatever observable differences in behavior there are, do the relevant specs actually allow for them, or could a fix be improved interoperability on these issues instead of allowing flexibility?

**Comment by xhwang-chromium** [comment]
Scope:

This is about the case where we run out of decoded video frames. It’s not related to network data. This affects all playbacks with audio AND video, for both MSE and SRC, with and without EME. This does not affect audio-only (where there’s no video) and video-only (where we don’t apply this logic) playbacks.

Context:

Previously we would trigger the pipeline to underflow immediately after we run out of decoded video frames. This caused some issues because relatively small video decoding delay (e.g. JS or device becomes busy) would pause-then-resume the playback and cause an audio glitch. This is much more noticeable to users than some dropped video frames. So we added the delay to alleviate this issue and only declare underflow when we run out of video frames for a certain period of time [1].

An alternative solution is to buffer more video frames in our media pipeline, but that could mean slower start-to-play time and more memory usage.

Overall, I feel this is an implementation quality issue, not a spec compliance issue. The 3-sec choice is a bit arbitrary, but the idea of having some grace period seems still valid to me.

Reference:
[1] Original CL: https://codereview.chromium.org/470773002/

**Comment by ddorwin** [comment]
>Whatever observable differences in behavior there are, do the relevant specs actually allow for them, or could a fix be improved interoperability on these issues instead of allowing flexibility?

I cannot say for certain that this is allowed by HTML and/or MSE, but for non-encrypted content it would seem to be okay assuming Chrome also reports the video frames as dropped. (This may also get into the discussion of whether `readyState` reflects network or pipeline status as mentioned in https://github.com/w3c/encrypted-media/issues/129#issuecomment-159462536.)

After discussing this, I'm not sure such behavior makes sense when a needed key is missing. First,
it seems extremely unlikely that such a glitch would be caused by a key that was delivered just a bit too late (when playback has already started). (However, theoretically, a video stream encrypted with alternating keys or with every other frame unencrypted could play back at half the frame rate in such a scenario.) Second, EME does not consider/allow dropping frames when a key is unavailable. Third, resuming video decoding at a future point may be more complicated since it's possible that some of those skipped encrypted blocks are necessary to decode the new current time.

Regardless of whether such a delay is allowed, the `waitingforkey` event should be fired at about the same time (vs., for example, three seconds before playback stops).

**Comment by xhwang-chromium** [comment]
I agree with ddorwin's analysis. I think the current EME spec makes sense on this and Chrome's current behavior isn't spec compliant.

I filed a chromium bug to track this: https://bugs.chromium.org/p/chromium/issues/detail?id=658422

**Comment by ddorwin** [comment]
tl;dr: The answer to the original question is "no," at least not such that it accommodates the current Chrome implementation.
